sdk/src/azure/core/az_log.c

```diff
@@ -16,75 +16,51 @@
-// false indicating whether it was logged.
-static bool _az_log_write_engine(bool log_it, az_log_classification classification, az_span message)
+void az_log_set_filter_callback(az_log_should_write_fn az_log_should_write_callback)
 {
-  // Copy the volatile fields to local variables so that they don't change within this function
-  az_log_message_fn const callback = _az_log_message_callback;
-  az_log_classification const* classifications = _az_log_classifications;
-
-  if (callback == NULL)
-  {
-    // If no one is listening, don't attempt to log.
-    return false;
-  }
-
-  if (classifications == NULL)
-  {
-    // If the user hasn't registered any classifications, then we log everything.
-    classifications
-        = &classification; // We don't need AZ_LOG_END_OF_LIST to be there, as very first comparison
-                           // is going to succeed and return from the function.
-  }
-
-  for (az_log_classification const* cls = classifications; *cls != AZ_LOG_END_OF_LIST; ++cls)
-  {
-    // If this message's classification is in the customer-provided list, we should log it.
-    if (*cls == classification)
-    {
-      if (log_it)
-      {
-        callback(classification, message);
-      }
-
-      return true;
-    }
-  }
-
-  // This message's classification is not in the customer-provided list; we should not log it.
-  return false;
+  _az_log_should_write_callback = az_log_should_write_callback;
 }
 
 // This function returns whether or not the passed-in message should be logged.
 bool _az_log_should_write(az_log_classification classification)
 {
-  return _az_log_write_engine(false, classification, AZ_SPAN_EMPTY);
+  _az_PRECONDITION(classification > 0);
+
+  // Copy the volatile fields to local variables so that they don't change within this function.
+  az_log_message_fn const message_callback = _az_log_message_callback;
+  az_log_should_write_fn const should_write_callback = _az_log_should_write_callback;
+
+  // If the user hasn't registered a should_write_callback, then we log everything, as long as a
+  // message_callback metho was provided.
+  return (should_write_callback == NULL || should_write_callback(classification))
+      && message_callback != NULL;
 }
 
 // This function attempts to log the passed-in message.
 void _az_log_write(az_log_classification classification, az_span message)
 {
-  (void)_az_log_write_engine(true, classification, message);
+  _az_PRECONDITION(classification > 0);
+  _az_PRECONDITION_VALID_SPAN(message, 0, true);
+
+  // Copy the volatile fields to local variables so that they don't change within this function.
+  az_log_message_fn const message_callback = _az_log_message_callback;
+
+  // If the user hasn't registered a should_write_callback, then we log everything, as long as a
+  // message_callback metho was provided.
+  if (_az_log_should_write(classification))
+  {
+    message_callback(classification, message);
```

Comment:
Make _az_log_should_write return the az_log_message_fn, instead of bool. This is the advantage we had with the "engine" function: we check and write, and when we do that, message_callback pointer has the same value in all points.
Here, you get the callback, it is NULL, other thread sets it to non-NULL, _az_log_should_write() returns you true, and you invoke a NULL callback.